complex_envs/visualize.py

```diff
@@ -0,0 +1,116 @@
+import numpy as np
+from scipy.integrate import simps
+import matplotlib.pyplot as plt
+from matplotlib import animation
+import seaborn as sns
+import yaml
+import argparse
+import os
+sns.set_style('white')
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-i', dest='rollout_file',
+        help='Path to the npz outputfile of a rollout',default='', type=str)
+    parser.add_argument('-f',dest='output_prefix',
+        help='Output visualization filenames start with this',default='output',type=str)
+    args = parser.parse_args()
+
+    #load in the numpy data
+    data = np.load(args.rollout_file)
+
+    #unpack the data
+    u_array = data['u_array']
+    impulse_array = data['impulse_array']
+    energy_array = data['energy_array']
+    x_mesh = data['x_mesh']
+    code_array = data['code_array']
+
+    #animation of demo
+    x_min = x_mesh[0]
+    x_max = x_mesh[-1]
+
+    y_min = min([np.min(u_array),np.min(impulse_array)])
+    y_max = max([np.max(u_array),np.max(impulse_array)])
+    frame_num = np.shape(u_array)[1]
+
+    fig = plt.figure()
+    ax1 = plt.axes(xlim=(x_min, x_max), ylim=(y_min,y_max))
+    line, = ax1.plot([], [], lw=2)
+    plt.xlabel('Bridge Position',fontsize=15)
+    plt.ylabel('Height',fontsize=15)
+
+    plotlays, plotcols = [3], ["red","black","green"]
+    plotlegends = ['bridge','','impulse']
+    lines = []
+    for index in range(3):
+        lobj = ax1.plot([],[],lw=2,color=plotcols[index],label=plotlegends[index])[0]
+        lines.append(lobj)
+    ax1.legend(loc='upper left')
+
+    def init():
+        for line in lines:
+            line.set_data([],[])
+        return lines
+
+
+
+
+    def animate(i):
```

Comment:
functions should be defined elsewhere and imported

Reply:
This is actually how the official matplotlib documentation recommends that you do this kind of animation workflow.  I agree it looks kind of messy to use function closures inside the main routine, but the alternative is defining a custom class that does exactly the same thing and calling it.

Reply:
I'll leave a comment linking to the official matplotlib documentation.